src/adaptors/tcp_adaptor.c

```diff
@@ -984,21 +997,24 @@ static qdr_tcp_connection_t *qdr_tcp_connection_ingress(qd_tcp_listener_t* liste
 
     tc->pn_raw_conn = pn_raw_connection();
     pn_raw_connection_set_context(tc->pn_raw_conn, tc);
+
+    qd_log(tcp_adaptor->log_source, QD_LOG_DEBUG, "[C%"PRIu64"] call pn_listener_raw_accept()", tc->conn_id);
+
     //the following call will cause a PN_RAW_CONNECTION_CONNECTED
     //event on another thread, which is where the rest of the
     //initialisation will happen, through a call to
     //qdr_tcp_connection_ingress_accept
-    qd_log(tcp_adaptor->log_source, QD_LOG_INFO, "[C%"PRIu64"] call pn_listener_raw_accept()", tc->conn_id);
-    pn_listener_raw_accept(listener->pn_listener, tc->pn_raw_conn);
-
-    const pn_netaddr_t *na = pn_raw_connection_remote_addr(tc->pn_raw_conn);
-    char remote_host[200];
-    char remote_port[50];
-    if (pn_netaddr_host_port(na, remote_host, 200, remote_port, 50) == 0) {
-        plog_set_string(tc->plog, PLOG_ATTRIBUTE_SOURCE_HOST, remote_host);
-        plog_set_string(tc->plog, PLOG_ATTRIBUTE_SOURCE_PORT, remote_port);
-    }
 
+    // IMPORTANT NOTE: this next call TO pn_listener_raw_accept may immediately schedule the connection on another I/O
+    // thread. IF you want to access tc after this call, the activation_lock  must be held to prevent the code in PN_RAW_CONNECTION_DISCONNECTED handler from running
+    // and freeing this connection(qdr_tcp_connection_t) from underneath. The PN_RAW_CONNECTION_DISCONNECTED event might occur if the
+    // client that initiated the connection, closes the connection while we are trying to accept the connection.
+    // For example:
+    //     LOCK(tc->activation_lock);
+    //     pn_listener_raw_accept(listener->pn_listener, tc->pn_raw_conn);
+    //     call_some_function(tc->conn_id);   // if the activation_lock is not held, tc might be freed
+    //     UNLOCK(tc->activation_lock);
+    pn_listener_raw_accept(listener->pn_listener, tc->pn_raw_conn);
     return tc;
```

Comment:
Not part of this PR - but probably deserves a separate issue: given the comment (which is great btw) then it is not safe to return tc to the caller, since the memory pointed to by tc can now be invalid (freed via another thread handling PR_RAW_CONNECTION_DISCONNECTED asynchronously).

Reply:
I added a bit of code to clarify the return values. Please take a look. I also renamed dispatcher to dispatcher_conn